sympy/solvers/polysys.py

```diff
@@ -429,3 +434,408 @@ def solve_triangulated(polys, *gens, **args):
+          must be satisfied for any solution to exist
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, Poly, ZZ
+    >>> from sympy.solvers.polysys import factor_system_poly
+    >>> a, b, c, x = symbols('a b c x')
+    >>> p1 = Poly((a - 1)*(x - 2), x, domain=ZZ[a,b,c])
+    >>> p2 = Poly((b - 3)*(x - 2), x, domain=ZZ[a,b,c])
+    >>> p3 = Poly(c, x, domain=ZZ[a,b,c])
+
+    The equation to be solved for x is ``x - 2 = 0`` provided either
+    of the two conditions on the parameters ``a`` and ``b`` is nonzero
+    and the constant parameter ``c`` should be zero.
+
+    >>> systems, constant_conds = factor_system_poly([p1, p2, p3])
+    >>> systems
+    [[(Poly(x - 2, x, domain='ZZ[a,b,c]'), (Poly(a - 1, x, domain='ZZ[a,b,c]'), Poly(b - 3, x, domain='ZZ[a,b,c]')))]]
+
+    >>> constant_conds
+    [Poly(c, x, domain='ZZ[a,b,c]')]
+
+    Notes
+    =====
+
+    - This is the core routine used by higher-level functions factor_system_cond,
+      factor_poly_system, and factor_system_bool
+    - Returns empty systems list ([]) when no solution exists
+    - Returns [[]] (list containing empty list) when any value is a solution
+    """
+    if not all(isinstance(poly, Poly) for poly in polys):
+        raise TypeError("polys should be a list of Poly instances")
+    if not polys:
+        return [[]], []
+    else:
+        base_domain = polys[0].domain
+        base_gens = polys[0].gens
+        if not all(poly.domain == base_domain and poly.gens == base_gens for poly in polys[1:]):
+            raise DomainError("All polynomials must have the same domain and generators")
+
+    constant_eqs = []
+    eqs_factors = []
+    conds_factor = defaultdict(list)
+
+    for poly in polys:
+        constant, factors_mult = poly.factor_list()
+
+        factors = [f for f, m in factors_mult]
+
+        if factors:
+            eqs_factors.append(factors)
+            if constant.is_zero is not False:
+                constp = constant.as_poly(base_gens, domain=base_domain)
+                for f in factors:
+                    if constp not in conds_factor[f]:
+                        conds_factor[f].append(constp)
+        elif constant.is_zero is True:
+            pass
+        elif constant.is_zero is False:
+            return ([], [])
+        else:
+            constp = constant.as_poly(base_gens, domain=base_domain)
+            constant_eqs.append(constp)
+
+    fac2conds = {f: tuple(conds) for f, conds in conds_factor.items()}
+
+    cnf = [[(f, fac2conds.get(f, ())) for f in eq] for eq in eqs_factors]
+
+    dnf = _cnf2dnf(cnf)
+
+    return dnf, constant_eqs
+
+
+def has_common_variables(expr1, expr2):
+    """Helper function to check if two expressions share any variables"""
+    try:
+        vars1 = expr1.free_symbols
+        vars2 = expr2.free_symbols
+        return bool(vars1.intersection(vars2))
+    except AttributeError:  # For non-symbolic expressions (e.g., integers)
+        return expr1 == expr2
```

Comment:
Don't catch `AttributeErrror`. What is this doing?

I think this should be a private function with a leading underscore in the name.

Reply:
How we originally implemented `_cnf2dnf`, well I refactored that to return ordered outputs but while writing back the docstrings I realised that the code accesses elements like `x[0]` and `y[0]` which aren't always subscriptable(like `Add`, `Symbol`, `int`) hence throwing `TypeError`. 

So I replaced the direct subscript access for which I wrote the helper function `has_common_variables`. Now since it uses `free_symbols` to handle symbolic expressions, it can throw an `AttributeError` in case of objects which don't take `free_symbols` as an attribute (eg. Tuples or integers).

Reply:
Why are there objects that don't have `free_symbols` being passed into this function?

Reply:
It's really because of how we are structuring the data in `factro_system_poly`'s base routine.  Notice the `cnf` variable has tuples of `(factors, conditions)`  :

```python
cnf = [[(f, fac2conds.get(f, ())) for f in eq] for eq in eqs_factors]
```

which is why we're getting non-symbolic objects in `_has_common_variables`.

The current approach catches the `AttributeError` and even though it works I do agree it's not ideal as it masks the underlying data structure issue.

Another approach might be to make the function explicitly handle tuple structure maybe ?
Something like:

```python
def _has_common_variables(expr1, expr2):
    """Helper function to check if two expressions share any variables"""
    if isinstance(expr1, tuple):
        expr1 = expr1[0]
    if isinstance(expr2, tuple):
        expr2 = expr2[0]

    vars1 = expr1.free_symbols
    vars2 = expr2.free_symbols
    return bool(vars1.intersection(vars2))
```

Reply:
The right approach is that the function should be called with the right type of arguments. There should be no need to check `isinstance` or to catch exceptions.

There is only one place in the code that actually calls this function so the function should be able to know exactly what type of input it receives.